### PR TITLE
Add ADR describing reasoning why we build images and security of it

### PR DESCRIPTION
**Pull request by potiuk**
The ADRs document the decision why Docker imaages are used as common
environment for CI and development environment, and also why building
images should be done in a secure way.

---
**^ Add meaningful description above**

Read the **[Pull Request Guidelines](https://github.com/apache/airflow/blob/main/CONTRIBUTING.rst#pull-request-guidelines)** for more information.
In case of fundamental code change, Airflow Improvement Proposal ([AIP](https://cwiki.apache.org/confluence/display/AIRFLOW/Airflow+Improvements+Proposals)) is needed.
In case of a new dependency, check compliance with the [ASF 3rd Party License Policy](https://www.apache.org/legal/resolved.html#category-x).
In case of backwards incompatible changes please leave a note in [UPDATING.md](https://github.com/apache/airflow/blob/main/UPDATING.md).

### CODE REVIEW

**Comment by potiuk** [comment]
cc: @bowrna @edithturn  @eladkal  @xurro - those are the next ADRs describing decisions behind why we are using docker images for our CI/developement as part of some clarifications and answering some questions from @Bowrna  and part of our 

I tried to capture two things:
* why we really need the images and the "smaller" projects approach with virtualenv won't work
* what are the decisions around security and building the images in a separate workflow  @ashb It woudl be great if you take a look if got all security aspects right.

File: dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md
Comment (eladkal) [inline dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md]:
```suggestion
* there is a "pull request target" workflow that uses only the code present in the protected "main" version
```

File: dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md
Comment (eladkal) [inline dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md:108]: Can this be enforced? Should we implement some kind of verification for this? (rather not relay on remembering)

**Comment by edithturn** [comment]
@potiuk. From what I see our CI workflows, not only involve Python dependencies but bash script(which we are trying to change), tests, database, document validation. I understand that to build all dependencies only in Python virtualenv enough, for the CI of other components we need docker to replicate the real environment in small containers that are also reusable.

Here is my question, if we have docker containers for airflow or breeze, then these are displayed on the GitHub Actions environment? Where the build and test take place. For these ADRs, what images are you referring to? :) 
Thanks in advance

**Comment by potiuk** [comment]
> @potiuk. From what I see our CI workflows, not only involve Python dependencies but bash script(which we are trying to change), tests, database, document validation. I understand that to build all dependencies only in Python virtualenv enough, for the CI of other components we need docker to replicate the real environment in small containers that are also reusable.

For the smaller scripts we will not need docker, virtualenv will be enough. The thing is that our small python scripts will only use "small" set of dependencies which are ok to have their own virtualenv. Those scripts will come from `dev` folder and we make sure that the scripts are taken from`main` and not from PR.  So we will not use Docker for those. 

For example you can see here:

https://github.com/apache/airflow/blob/4ac35d723b73d02875d56bf000aafd2235ef0f4a/.github/workflows/ci.yml#L240

This is the action that is runs pytest tests of our "Breeze2".

What it does:
* it sets working directory to ".dev/breeze"
* it sets-up python in version 3.7 with dependency cache that understands pip install
* It installs dependencies specified in "./dev/breeze"' setup.* files
* and runs tests afterwards

This will be very similar pattern for most of our jobs that do not need full "550" dependencies of airflow installed. In our case the CI image is really needed to run airflow tests, to run mypy and flake and few other things (for example building providers - which also needs to be done inside the container for security).

Most of the other scripts of ours (free space for example) are perfectly fine to use the virtualenv support provided by GithubActions like below: 

```
  run-new-breeze-tests:
    timeout-minutes: 10
    name: Breeze2 tests
    runs-on: ${{ fromJson(needs.build-info.outputs.runsOn) }}
    needs: [build-info]
    defaults:
      run:
        shell: bash
        working-directory: ./dev/breeze
    steps:
      - uses: actions/checkout@v2
        with:
          persist-credentials: false
      - uses: actions/setup-python@v2
        with:
          python-version: '3.7'
          cache: 'pip'
      - run: pip install -e .
      - run: python3 -m pytest -n auto --color=yes
```

**Comment by potiuk** [comment]
> Here is my question, if we have docker containers for airflow or breeze, then these are displayed on the GitHub Actions environment? Where the build and test take place. For these ADRs, what images are you referring to? :)
> Thanks in advance

All the details on where the images are build and set are described in https://github.com/apache/airflow/blob/main/CI.rst#ci-sequence-diagrams . In short we have two workflows (in https://github.com/apache/airflow/tree/main/.github/workflows) :

* build-images.yml - > this is the "pull_request_target" workflow that builds and pushes the images to "ghcr.io" (write access)
* ci.yml - > this is the "pull request" workflow that executes tests, provider build, etc. etc. - some of the jobs there require the "CI images" and they are pulled before tests are executed for example. This is "read-only" workflow.

File: dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md
Comment (potiuk) [reply dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md:108]: That's a good point. We could make a pre-commit check for it likely.

**Review by eladkal — APPROVED** [review]

**Comment by potiuk** [comment]
@uranusjr and @ashb as you were involved in the CI scripting comments and the ADR's are mainly for the future/reasoning I'd also love your comments before merging that one. 

I will keep on updating the ADRs as we progress with the implementation of the Breeze rewrite and more questions will be asked by @edithturn and @Bowrna  - this is actually  agreat opportunity to capture all the reasoning behind many of the decisions made :D.

File: dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md
Comment (potiuk) [reply dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md:108]: https://github.com/apache/airflow/pull/20430

Comment (potiuk) [reply dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md:108]: PR ready @eladkal :)

Comment (eladkal) [reply dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md:108]: thanks!

**Comment by potiuk** [comment]
Merging for now - those ADRs are still Draft, so we can always modify them later.

File: dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md
Comment (Bowrna) [inline dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md:50]: @potiuk could you explain me what jobs mean in this case?

File: dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md
Comment (Bowrna) [inline dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md:142]: Under any circumstances any of the scripts coming from the outside of `scripts/ci` and `dev` should be
   executed on the host during the image building process.

Is the above line correct?

File: dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md
Comment (potiuk) [reply dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md:50]: Yep. You can see the jobs in .github/workflows/ci.yml  - and when you create PR you will see the red/green status in 'checks' tab - every such red/green stuff is a job. For example 'tests postres python 3.7' - is an example of a job.

Comment (potiuk) [reply dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md:50]: https://docs.github.com/en/actions/learn-github-actions/workflow-syntax-for-github-actions#jobs

File: dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md
Comment (potiuk) [reply dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md:142]: Yep. This is advanced security stuff for our 'build' jobs. We really need to make sure that we only execute anything in the build.yml that is coming from one of those two folders only. Because all the rest can be modified by the users making PR and they might get access to stuff we do not want them to.

What we do in our workflows, we make sure that those two folders (for now only scripts/ci) are not coming from the PR of the user but from 'main' branch of our repository

Comment (Bowrna) [reply dev/breeze/doc/adr/0005-preventing-using-contributed-code-when-building-images.md:142]: @potiuk Under any circumstances any of the scripts coming from the outside of scripts/ci and dev should **not** be executed on the host during the image building process.
Am i right? Not is missing, which will mean you allow any scripts to be executed

**Pull request merged**
